data/transactions/verify/txn.go

```diff
@@ -314,3 +314,113 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
 	return nil
 
 }
+
+// PaysetGroups verifies that the payset have a good signature and that the underlying
+// transactions are properly constructed.
+// Note that this does not check whether a payset is valid against the ledger:
+// a PaysetGroups may be well-formed, but a payset might contain an overspend.
+//
+// This version of verify is performing the verification over the provided execution pool.
+func PaysetGroups(ctx context.Context, payset [][]transactions.SignedTxn, blk bookkeeping.Block, verificationPool execpool.BacklogPool) (err error) {
+	txnGroupIdx := 0
+	const txnPerGroupThreshold = 32
+	proto, ok := config.Consensus[blk.BlockHeader.CurrentProtocol]
+	if !ok {
+		return protocol.Error(blk.BlockHeader.CurrentProtocol)
+	}
+	if len(payset) == 0 {
+		return nil
+	}
+	spec := transactions.SpecialAddresses{
+		FeeSink:     blk.BlockHeader.FeeSink,
+		RewardsPool: blk.BlockHeader.RewardsPool,
+	}
+	var txnGroups [][]transactions.SignedTxn
+	// prepare up to 16 concurrent worksets.
+	worksets := make(chan struct{}, 16)
+	worksDoneCh := make(chan interface{}, 16)
+	zeroAddress := basics.Address{}
+	processing := 0
+	tasksCtx, cancelTasksCtx := context.WithCancel(ctx)
+	defer cancelTasksCtx()
+	for processing >= 0 {
+		if txnGroupIdx < len(payset) && (len(txnGroups) == 0) {
+			txnCounter := 0 // how many transaction we already included in the current workset.
+			for i := txnGroupIdx; i < len(payset); i++ {
+				if txnCounter+len(payset[i]) > txnPerGroupThreshold {
+					if i == txnGroupIdx {
+						i++
+					}
+					txnGroups = payset[txnGroupIdx:i]
+					txnGroupIdx = i
+					break
+				}
+				if i == len(payset)-1 {
+					txnGroups = payset[txnGroupIdx:]
+					txnGroupIdx = len(payset)
+					break
+				}
+				txnCounter += len(payset[i])
+			}
```

Comment:
possibly needs to be a method

Reply:
I think it would make it quite confusing to have it as a separate function, given that it would need to be something like:
```
func getNextWorkset(payset [][]transactions.SignedTxn, txnGroupIdx *int) (txnGroups [][]transactions.SignedTxn) {
                        txnCounter := 0 // how many transaction we already included in the current workset.
			// scan starting from the current position until we filled up the workset.
			for i := *txnGroupIdx; i < len(payset); i++ {
				if txnCounter+len(payset[i]) > txnPerWorksetThreshold {
					if i == *txnGroupIdx {
						i++
					}
					txnGroups = payset[txnGroupIdx:i]
					*txnGroupIdx = i
					return
				}
				if i == len(payset)-1 {
					txnGroups = payset[*txnGroupIdx:]
					*txnGroupIdx = len(payset)
					return
				}
				txnCounter += len(payset[i])
			}
                        return nil
} 
```

Reply:
or
```
type workset struct {
  payset [][]transactions.SignedTxn
  idx int
}

func (w *workset) next() (txnGroups [][]transactions.SignedTxn) {
...
}

func (w *workset) completed() bool {
  return w.idx >= len(w.payset)
}
```

Reply:
that looks much better; I'll add that.